site/en/docs/android/trusted-web-activity/offline-first/index.md

```diff
@@ -0,0 +1,197 @@
+    // If TWA has already launched successfully, launch TWA immediately.
+    // Otherwise, check connection status. If online, launch the Trusted Web Activity with `launchTwa()`.
+    // Otherwise, if offline, render the offline fallback screen.
+    if (hasTwaLaunchedSuccessfully()) {
+        launchTwa();
+    } else if (isOnline()) {
+        firstTimeLaunchTwa();
+    } else {
+        renderOfflineFallback();
+    }
+}
+```
+
+The previous code handles three scenarios:
+
+- If the TWA has launched previously, this means that the service worker has
+  been registered, and the PWA will be able to respond offline. In that case,
+  call the method `launchTwa()` defined in the parent class to launch the
+  Trusted Web Activity directly.
+- If the TWA hasn't launched previously and the user is online, launch the
+  Trusted Web Activity for the first time using the `firstTimeLaunchTwa()`
+  method that you'll implement later.
+- If the TWA hasn't already launched and the user is offline, render the native
+  offline fallback screen.
+
+## Implement helper methods
+
+The final step is to implement the helper methods called by the previous code.
+Here's how the code to check of offline state `isOnline()` looks like:
+
+```java
+private boolean isOnline() {
+    ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
+    NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
+    return activeNetworkInfo != null && activeNetworkInfo.isConnected();
+}
+```
+
+Next, implement `hasTwaLaunchedSuccessfully()` used to check if the TWA has
+launched at least once:
+
+```java
+private boolean hasTwaLaunchedSuccessfully() {
+    // Return `true` if the preference "twa_launched_successfully" has already been set.
+    SharedPreferences sharedPref = getSharedPreferences(getString(R.string.twa_offline_first_preferences_file_key), Context.MODE_PRIVATE);
+    return sharedPref.getBoolean(getString(R.string.twa_launched_successfully), false);
+}
+```
+
+The previous code calls the `launchTWA()` from the parent class, and saves the
+`twa_launched_successfully` flag in shared preferences, to indicate that the TWA
+has launched successfully, at least once.
+
+The remaining helper method: `renderOfflineFallback()` renders an Android
+offline screen. 
+
+```java
+private void renderOfflineFallback() {
+    setContentView(R.layout.activity_offline_first_twa);
+
+    Button retryBtn = this.findViewById(R.id.retry_btn);
+    retryBtn.setOnClickListener(new View.OnClickListener() {
+        public void onClick(View v) {
+            // Check connection status. If online, launch the Trusted Web Activity for the first time.
+            if (isOnline()) firstTimeLaunchTwa();
+        }
+    });
+}
+```
+
+Fort his demo, we have defined the `activity_offline_first_twa` layout, which
+contains a button to retry, which will, in time, to execute
+`firstTimeLaunchTwa()` after checking connection:
```

Comment:
```suggestion
`firstTimeLaunchTwa()` after checking the connection.
```

Reply:
Colons aren't used to introduce images. If you feel that the image needs to be referenced in text, add something like, 'see the image below'.